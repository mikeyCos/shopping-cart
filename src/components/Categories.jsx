import PropTypes from "prop-types";
import NavAnchor from "./NavAnchor";
import styles from "../styles/Categories.module.css";

const Categories = ({ categories }) => {
  return (
    <article className={styles["categories"]}>
      <ul className={styles["categories-list"]}>
        {categories.map((category) => {
          return (
            <li key={category}>
              <NavAnchor
                pathname={`category/${category.replaceAll(" ", "-")}`}
                text={category}
              />
            </li>
          );
        })}
      </ul>
    </article>
  );
};

Categories.propTypes = {
  categories: PropTypes.array,
};

export default Categories;
