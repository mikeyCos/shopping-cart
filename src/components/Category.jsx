import {
  Link,
  useLocation,
  useOutletContext,
  useParams,
} from "react-router-dom";
import parseCategory from "../utilities/parseCategory";
import styles from "../styles/Category.module.css";

/* Renders products based on the useParams category value
 * useParams category value will fallback to 'all'
 */
const Category = () => {
  const { category = "all" } = useParams();
  const [products] = useOutletContext();
  const { pathname } = useLocation();
  const productKey = parseCategory(category);
  const headingCategory = category.replaceAll("-", " ");
  // console.log(products);
  // console.log("category", category);
  // console.log("productKey", productKey);
  return (
    <section>
      <h2>Category</h2>
      <h3>{headingCategory}</h3>
      {products && (
        <section className={styles.products} role="region">
          {products[productKey].map((product) => (
            <article className={styles["product-card"]} key={product.id}>
              <Link
                to={`product/${encodeURIComponent(product.title)}`}
                state={{ product, previousLocation: pathname }}
              >
                <picture>
                  <img
                    className={styles["product-img"]}
                    src={product.image}
                    loading="lazy"
                    alt="#"
                  />
                </picture>
              </Link>

              <div className="info">
                <h4>{product.title}</h4>
                <p>${product.price}</p>
              </div>
            </article>
          ))}
        </section>
      )}
    </section>
  );
};

export default Category;
