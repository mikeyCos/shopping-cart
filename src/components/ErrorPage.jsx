import { useRouteError } from "react-router-dom";
import styles from "../styles/ErrorPage.module.css";

const ErrorPage = () => {
  const { error, status } = useRouteError();

  return (
    <section id={styles["error-section"]} role="region">
      {status && <h1>{status}</h1>}
      <p>{error.message}</p>
    </section>
  );
};

export default ErrorPage;
