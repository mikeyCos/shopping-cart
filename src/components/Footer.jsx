import "../styles/Footer.module.css";

const Footer = () => {
  return (
    <footer>
      <div>Placeholder</div>
    </footer>
  );
};

export default Footer;
