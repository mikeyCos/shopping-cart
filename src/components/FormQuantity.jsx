import PropTypes from "prop-types";
import styles from "../styles/FormQuantity.module.css";

const FormQuantity = ({ submitForm, children }) => {
  const onSubmitHandler = (e) => {
    e.preventDefault();
    submitForm();
  };

  return (
    <form
      noValidate={true}
      onSubmit={onSubmitHandler}
      className={styles["form-quantity"]}
    >
      {children}
    </form>
  );
};

const SubmitButton = ({ text }) => (
  <button type="submit" className={styles["submit-btn"]}>
    {text}
  </button>
);

FormQuantity.SubmitButton = SubmitButton;

FormQuantity.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.object,
    PropTypes.arrayOf(PropTypes.object),
  ]),
  submitForm: PropTypes.func.isRequired,
};

SubmitButton.propTypes = {
  text: PropTypes.string.isRequired,
};

export default FormQuantity;

/* 
// Product.jsx
<form noValidate={true} onSubmit={onSubmitHandler}>
  <p>{product.description}</p>
  <p>{formatPrice(product.price)}</p>
  <InputQuantity
    quantity={quantity}
    setQuantity={onChangeQuantityHandler}
    incrementHandler={incrementQuantityHandler}
    decrementHandler={decrementQuantityHandler}
  />
  <button type="submit">Add to cart</button>
</form>; */

/* 
// Cart.jsx
<form
  noValidate={true}
  onSubmit={(e) => {
    e.preventDefault();
    removeFromCart(item.id);
  }}
>
  <p>{formatPrice(item.price)}</p>
  <InputQuantity
    quantity={item.quantity}
    setQuantity={(newQuantity) => {
      console.log("previous quantity", item.quantity);
      setQuantity(item.id, newQuantity);
    }}
    incrementHandler={() => {
      const newQuantity = +item.quantity + 1;
      newQuantity <= 999 && setQuantity(item.id, newQuantity);
    }}
    decrementHandler={() => {
      const newQuantity = item.quantity - 1;
      newQuantity >= 0 && setQuantity(item.id, newQuantity);
    }}
  />
  <button type="submit">Delete</button>
</form>; */
