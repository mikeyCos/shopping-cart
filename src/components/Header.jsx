import { Link } from "react-router-dom";
import GitHubIcon from "../assets/icons/github-mark/github-mark.svg?react";
import "../styles/Header.css";

const Header = () => {
  return (
    <header>
      <nav>
        <ul className="nav-left">
          <li>
            <Link to="/">
              <img src="#" alt="" />
              <h1>Logo</h1>
            </Link>
          </li>
        </ul>

        <ul className="nav-right">
          <li>
            <Link to="/Placeholder_00">Placeholder_00</Link>
          </li>

          <li>
            <Link to="/Placeholder_01">Placeholder_01</Link>
          </li>

          <li>
            <a href="#" target="_blank">
              <GitHubIcon className="icon github" />
            </a>
          </li>
        </ul>
      </nav>
    </header>
  );
};

export default Header;
