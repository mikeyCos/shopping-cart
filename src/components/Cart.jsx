import PropTypes from "prop-types";
import { createContext, useContext, useState } from "react";
import { Link } from "react-router-dom";
import ProductCard from "./ProductCard";
import FormQuantity from "./FormQuantity";
import InputQuantity from "./InputQuantity";
import formatPrice from "../utilities/formatPrice";
import styles from "../styles/Cart.module.css";

const Cart = () => {
  const { cart, removeFromCart, setQuantity } = useContext(CartContext);

  const subTotal = cart.reduce((accumulator, currentCartItem) => {
    return accumulator + currentCartItem.price * currentCartItem.quantity;
  }, 0);

  return (
    <section id={styles["cart"]} role="region">
      Cart
      {cart.length === 0 && <h4>Cart is empty</h4>}
      {cart.length > 0 && (
        <section role="region" className={styles["cart-container"]}>
          {cart.map((item) => {
            return (
              <ProductCard
                key={item.id}
                className={styles["cart-product-card"]}
              >
                <ProductCard.Heading title={item.title} />
                <Link
                  to={`/shop/product/view/${encodeURIComponent(item.title)}`}
                  state={{ product: item }}
                >
                  <ProductCard.Picture
                    src={item.image}
                    alt="#"
                    className={styles["product-picture"]}
                  />
                </Link>
                <FormQuantity submitForm={() => removeFromCart(item.id)}>
                  <p>{formatPrice(item.price)}</p>
                  <div className="form-controls">
                    <InputQuantity
                      quantity={item.quantity}
                      setQuantity={(newQuantity) => {
                        console.log("previous quantity", item.quantity);
                        setQuantity(item.id, newQuantity);
                      }}
                      incrementHandler={() => {
                        const newQuantity = +item.quantity + 1;
                        newQuantity <= 999 && setQuantity(item.id, newQuantity);
                      }}
                      decrementHandler={() => {
                        const newQuantity = item.quantity - 1;
                        newQuantity >= 0 && setQuantity(item.id, newQuantity);
                      }}
                    />
                    <FormQuantity.SubmitButton text="Delete" />
                  </div>
                </FormQuantity>
              </ProductCard>
            );
          })}
          <p className={styles["subtotal"]}>
            Subtotal: <span>{formatPrice(subTotal)}</span>
          </p>
        </section>
      )}
    </section>
  );
};

/* React hook createContext
 * https://react.dev/learn/passing-data-deeply-with-context
 */
const CartContext = createContext([]);

const CartProvider = ({ children }) => {
  const [cart, setCart] = useState([]);
  /* Removing item from cart
   * Check if item is in cart
   * If item is in cart
   *  Reduce cart and add quantities of item
   * Else
   *  Add item to cart
   *
   * Update item's quantity
   */
  const addToCart = (product, quantity) => {
    const { id } = product;

    /* // Option 1:
    const newCart = productInCart
      ? cart.map((productInCart) => {
          if (productInCart.id === id)
            return {
              ...productInCart,
              quantity: productInCart.quantity + quantity,
            };
          return productInCart;
        })
      : [...cart, { ...product, quantity }];
    setCart(newCart); */

    let isInCart = false;
    const newCart = cart.map((productInCart) => {
      if (productInCart.id === id) {
        isInCart = true;
        return {
          ...productInCart,
          quantity: productInCart.quantity + quantity,
        };
      }

      return productInCart;
    });

    setCart(isInCart ? newCart : [...cart, { ...product, quantity }]);

    /* // Option 2:
    setCart((prevCart) => {
      const isInCart = prevCart.some((cartItem) => cartItem.id === product.id);

      return isInCart
        ? prevCart.reduce((accumulator, currentCartItem) => {
            return [
              ...accumulator,
              currentCartItem.id === product.id
                ? {
                    ...currentCartItem,
                    quantity: currentCartItem.quantity + quantity,
                  }
                : currentCartItem,
            ];
          }, [])
        : [...prevCart, { ...product, quantity }];
    }); */
  };

  const removeFromCart = (productID) => {
    setTimeout(() => {
      setCart((prevCart) => {
        return prevCart.filter((product) => product.id !== productID);
      });
    }, 250);
  };

  const setQuantity = (productID, quantity) => {
    if (quantity == 0) {
      removeFromCart(productID);
    } else {
      setCart((prevCart) =>
        prevCart.map((product) => {
          return product.id === productID ? { ...product, quantity } : product;
        })
      );
    }
  };

  return (
    <CartContext.Provider
      value={{ cart, addToCart, removeFromCart, setQuantity }}
    >
      {children}
    </CartContext.Provider>
  );
};

CartProvider.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
  ]).isRequired,
};

export { Cart as default, CartContext, CartProvider };
