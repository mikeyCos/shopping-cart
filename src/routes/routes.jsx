import { useRoutes } from "react-router-dom";
import BasicLayout from "../layouts/BasicLayout";
import Modal from "../components/Modal";
import Product from "../components/Product";
import ErrorPage from "../components/ErrorPage";
import Home from "../components/Home";
import Shop from "../components/Shop";
import Cart from "../components/Cart";
import Category from "../components/Category";

const defaultRoutes = [
  {
    element: <BasicLayout />,
    errorElement: <ErrorPage />,
    children: [
      {
        path: "/",
        element: <Home />,
      },
      {
        path: "/shop",
        element: <Shop />,
        children: [
          { index: true, element: <Category /> },
          { path: "category/:category", element: <Category /> },
        ],
      },
      {
        path: "/cart",
        element: <Cart />,
      },
    ],
  },
];

const productRoutes = [
  {
    element: <Modal />,
    path: "/shop",
    children: [
      { index: true, path: "product/:product", element: <Product /> },
      { path: "category/:category/product/:product", element: <Product /> },
    ],
  },
];

const DefaultRoutes = ({ location }) => {
  const routes = useRoutes(defaultRoutes, location);
  return routes;
};

const ProductRoutes = () => {
  const routes = useRoutes(productRoutes);
  return routes;
};

export { DefaultRoutes as default, ProductRoutes };
