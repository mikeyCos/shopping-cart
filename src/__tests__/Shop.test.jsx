import { beforeEach, describe, expect, it, vi } from "vitest";
import { render, screen } from "@testing-library/react";
import { RouterProvider, createMemoryRouter } from "react-router-dom";
import { fetchMock } from "./mocks";
import { CartProvider } from "../components/Cart";
import routes from "../routes/routes";

/* Optional
 * Define fetchMock here
 * expect(fetchMock).toBeCalledTimes
 */
/* const fetchMock = vi.fn(() => {
  json: vi.fn().mockReturnValueOnce(categories).mockReturnValueOnce(products);
});
vi.stubGlobal("fetch", fetchMock);
vi.mock("../utilities/parseProducts");
 */

beforeEach(() => {
  vi.mock("../utilities/parseProducts");
  window.scrollTo = vi.fn();
  fetchMock();
});

describe("Shop component", () => {
  it("The Shop component component matches snapshot", async () => {
    const router = createMemoryRouter(routes, { initialEntries: ["/shop"] });
    const { container } = render(
      <CartProvider>
        <RouterProvider router={router} />
      </CartProvider>
    );
    expect(container).toMatchSnapshot();
  });

  it("Renders correct text when fetch fails", async () => {
    const rejectCause = { status: 404 };
    window.fetch.mockRejectedValue(
      new Error("API is down", { cause: { ...rejectCause } })
    );
    const router = createMemoryRouter(routes, { initialEntries: ["/shop"] });

    render(
      <CartProvider>
        <RouterProvider router={router} />
      </CartProvider>
    );

    const errorMessage = await screen.findByText("API is down");
    expect(errorMessage).toBeInTheDocument();
  });

  it("The 'All' category is first in the categories list", async () => {
    const router = createMemoryRouter(routes, { initialEntries: ["/shop"] });
    render(
      <CartProvider>
        <RouterProvider router={router} />
      </CartProvider>
    );

    const lists = await screen.findAllByRole("list");
    const categoriesListItems = lists.find((list) =>
      list.textContent.match(/all/i)
    ).children;

    const firstCategory = categoriesListItems[0];
    expect(firstCategory.textContent).match(/all/i);
  });
});
