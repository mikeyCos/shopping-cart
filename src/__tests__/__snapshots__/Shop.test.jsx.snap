// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Shop component > The Shop component component matches snapshot 1`] = `<div />`;
